### ISSUE TITLE
[@typescript-eslint/no-unused-vars] false positive with optional chaining

### ISSUE THREAD

**Issue by bradennapier** [typescript-eslint/typescript-eslint#2473]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have read the FAQ and my problem is not listed.

Getting false positive `no-unused-vars` when using optional chaining

```typescript
[{}, { type: 1 }].map((m) => ({
    type: m?.type,
}))
```

Expected Result

No linting issues

Actual Result

```
'm' is defined but never used.eslint@typescript-eslint/no-unused-vars
```

Versions

package	version
@typescript-eslint/eslint-plugin	4.0.1
@typescript-eslint/parser	4.0.1
TypeScript	4.0.2
ESLint	7.8.0
node	12.14.1

**Comment by clement9527** [comment]
And also false positive on @typescript-eslint/no-unused-expressions

**Comment by bradzacher** [comment]
please file a new issue. this issue is completely unrelated to that rule.

**Comment by bradennapier** [comment]
> Could you please confirm to me here that your lint run is using the latest version of eslint?
> 
> Eslint core has full support for optional chaining, so this should be supported out of the box. [repl](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IFwibm8tdW51c2VkLXZhcnNcIjogW1wiZXJyb3JcIiwgeyBcInZhcnNcIjogXCJhbGxcIiwgXCJhcmdzXCI6IFwiYWxsXCIsIFwiaWdub3JlUmVzdFNpYmxpbmdzXCI6IGZhbHNlIH1dICovXG5cblt7fSwgeyB0eXBlOiAxIH1dLm1hcCgobSkgPT4gKHtcbiAgICB0eXBlOiBtPy50eXBlLFxufSkpIiwib3B0aW9ucyI6eyJwYXJzZXJPcHRpb25zIjp7ImVjbWFWZXJzaW9uIjoxMiwic291cmNlVHlwZSI6InNjcmlwdCIsImVjbWFGZWF0dXJlcyI6e319LCJydWxlcyI6eyJjb25zdHJ1Y3Rvci1zdXBlciI6MiwiZm9yLWRpcmVjdGlvbiI6MiwiZ2V0dGVyLXJldHVybiI6Miwibm8tYXN5bmMtcHJvbWlzZS1leGVjdXRvciI6Miwibm8tY2FzZS1kZWNsYXJhdGlvbnMiOjIsIm5vLWNsYXNzLWFzc2lnbiI6Miwibm8tY29tcGFyZS1uZWctemVybyI6Miwibm8tY29uZC1hc3NpZ24iOjIsIm5vLWNvbnN0LWFzc2lnbiI6Miwibm8tY29uc3RhbnQtY29uZGl0aW9uIjoyLCJuby1jb250cm9sLXJlZ2V4IjoyLCJuby1kZWJ1Z2dlciI6Miwibm8tZGVsZXRlLXZhciI6Miwibm8tZHVwZS1hcmdzIjoyLCJuby1kdXBlLWNsYXNzLW1lbWJlcnMiOjIsIm5vLWR1cGUtZWxzZS1pZiI6Miwibm8tZHVwZS1rZXlzIjoyLCJuby1kdXBsaWNhdGUtY2FzZSI6Miwibm8tZW1wdHkiOjIsIm5vLWVtcHR5LWNoYXJhY3Rlci1jbGFzcyI6Miwibm8tZW1wdHktcGF0dGVybiI6Miwibm8tZXgtYXNzaWduIjoyLCJuby1leHRyYS1ib29sZWFuLWNhc3QiOjIsIm5vLWV4dHJhLXNlbWkiOjIsIm5vLWZhbGx0aHJvdWdoIjoyLCJuby1mdW5jLWFzc2lnbiI6Miwibm8tZ2xvYmFsLWFzc2lnbiI6Miwibm8taW1wb3J0LWFzc2lnbiI6Miwibm8taW5uZXItZGVjbGFyYXRpb25zIjoyLCJuby1pbnZhbGlkLXJlZ2V4cCI6Miwibm8taXJyZWd1bGFyLXdoaXRlc3BhY2UiOjIsIm5vLW1pc2xlYWRpbmctY2hhcmFjdGVyLWNsYXNzIjoyLCJuby1taXhlZC1zcGFjZXMtYW5kLXRhYnMiOjIsIm5vLW5ldy1zeW1ib2wiOjIsIm5vLW9iai1jYWxscyI6Miwibm8tb2N0YWwiOjIsIm5vLXByb3RvdHlwZS1idWlsdGlucyI6Miwibm8tcmVkZWNsYXJlIjoyLCJuby1yZWdleC1zcGFjZXMiOjIsIm5vLXNlbGYtYXNzaWduIjoyLCJuby1zZXR0ZXItcmV0dXJuIjoyLCJuby1zaGFkb3ctcmVzdHJpY3RlZC1uYW1lcyI6Miwibm8tc3BhcnNlLWFycmF5cyI6Miwibm8tdGhpcy1iZWZvcmUtc3VwZXIiOjIsIm5vLXVuZGVmIjoyLCJuby11bmV4cGVjdGVkLW11bHRpbGluZSI6Miwibm8tdW5yZWFjaGFibGUiOjIsIm5vLXVuc2FmZS1maW5hbGx5IjoyLCJuby11bnNhZmUtbmVnYXRpb24iOjIsIm5vLXVudXNlZC1sYWJlbHMiOjIsIm5vLXVudXNlZC12YXJzIjoyLCJuby11c2VsZXNzLWNhdGNoIjoyLCJuby11c2VsZXNzLWVzY2FwZSI6Miwibm8td2l0aCI6MiwicmVxdWlyZS15aWVsZCI6MiwidXNlLWlzbmFuIjoyLCJ2YWxpZC10eXBlb2YiOjJ9LCJlbnYiOnt9fX0=)
> 
> Another user had issues because they had multiple versions of eslint installed

Yep, and just updated to `7.8.1` as well same issue.

**Comment by bradzacher** [comment]
I'm unable to repro this against either `master` or the `4.0.1` tag.
I think you might be running into an issue with multiple ESLint versions.

Could you please create a repro repo for me to help me look into this?

**Comment by bradennapier** [comment]
Yeah I think that I had a stuck version of eslint server running even through troubleshooting, restarts of VSCode, and days with this problem (and a few others which were upsetting, lol).  I just upgraded OSX today and in doing so restarted my computer for the first time in..... awhile .... :-P and it seems to be fine now so I figured VSCode maybe didn't properly terminate the process for whatever reason when i restarted one time then it just sat there literally being a bug... and i hate bugs... total phobia of the things.

Thanks for looking into it!
